### ISSUE TITLE
[Bug Report] fp32 dest accumulation unsupported for REDUCE_SCALAR

### ISSUE THREAD

**Issue by hschoi4448** [tenstorrent/tt-metal#8424]
**Describe the bug**
I am working on adding support for `fp32_dest_acc_en` to ops for training support.
However, I discovered that REDUCE_SCALAR does not support fp32_dest_acc_en.

**To Reproduce**
Please check the link below.
https://github.com/tenstorrent/tt-llk-wh-b0/blob/2b92e02e8e723f82e9c1a87049c96b140b103c7a/llk_lib/llk_math_reduce.h#L195

**Expected behavior**
I hope that `fp32_dest_acc_en` can be used for REDUCE_SCALAR.

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit )

**Additional context**

**Comment by razorback3** [comment]
@jliangTT 
Is it preferable for you to let us select a priority label on our own or notify you to select an appropriate priority label?

**Comment by davorchap** [comment]
> @jliangTT 
> 
> Is it preferable for you to let us select a priority label on our own or notify you to select an appropriate priority label?

@ttmtrajkovic can triage and assign

**Comment by dongjin-na** [comment]
FYI, the backward operation of scalar bias add also causes issues due to the use of REDUCE_SCALAR.
https://github.com/tenstorrent/tt-metal/blob/8788_linear_fp32/tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_single_core_hw.cpp

**Comment by tt-rkim** [comment]
@ttmtrajkovic @davorchap any updates on this?

**Comment by ttmtrajkovic** [comment]
hi @dongjin-na,
there's a hardware limitation to use fp32 accumulation with reduce_scalar and I will try to work around it, but it would be good to have a test / branch where I'd able to test it - could you provide one?

On another note, which REDUCE_OP are you running - max or average? With max reduce, you don't get anything by doing fp32 accumulation as inputs are fp16b, but if you're working with reduce average then there's benefit for sure.

Milos

**Comment by dongjin-na** [comment]
Dear @ttmtrajkovic,

The scalar bias add backward kernel uses `PoolType::SUM' and `ReduceDim::REDUCE_SCALAR'.

I've uploaded test code in the following branch: [8788_reduce_scalar](https://github.com/tenstorrent/tt-metal/tree/8788_reduce_scalar).

To run the tests, use the following command:
```
(python_env) ubuntu@tt-metal-dev-moreh-wh-9:~/tt-metal$ pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py::test_reduce_scalar_with_fp32_acc
===================================================================================================================== test session starts =====================================================================================================================
platform linux -- Python 3.8.10, pytest-7.2.2, pluggy-1.5.0 -- /home/ubuntu/tt-metal/python_env/bin/python3
cachedir: .pytest_cache
rootdir: /home/ubuntu/tt-metal, configfile: pytest.ini
plugins: dash-2.15.0, split-0.8.2, timeout-2.2.0, anyio-4.3.0
timeout: 2400.0s
timeout method: signal
timeout func_only: False
collected 1 item

tests/tt_eager/python_api_testing/unit_testing/misc/test_moreh_linear.py::test_reduce_scalar_with_fp32_acc[requires_grads0-shapes0]                  Device | INFO     | Opening user mode device driver
  Detecting chips (found 2)
2024-05-30 15:00:50.733 | INFO     | SiliconDriver   - Detected 1 PCI device : {0}
2024-05-30 15:00:50.745 | WARNING  | SiliconDriver   - init_detect_tt_device_numanodes(): Could not determine NumaNodeSet for TT device (physical_device_id: 0 pci_bus_id: 0000:07:00.0)
2024-05-30 15:00:50.745 | WARNING  | SiliconDriver   - Could not find NumaNodeSet for TT Device (physical_device_id: 0 pci_bus_id: 0000:07:00.0)
2024-05-30 15:00:50.747 | WARNING  | SiliconDriver   - bind_area_memory_nodeset(): Unable to determine TT Device to NumaNode mapping for physical_device_id: 0. Skipping membind.
---- ttSiliconDevice::init_hugepage: bind_area_to_memory_nodeset() failed (physical_device_id: 0 ch: 0). Hugepage allocation is not on NumaNode matching TT Device. Side-Effect is decreased Device->Host perf (Issue #893).
2024-05-30 15:00:50.750 | WARNING  | SiliconDriver   - bind_area_memory_nodeset(): Unable to determine TT Device to NumaNode mapping for physical_device_id: 0. Skipping membind.
---- ttSiliconDevice::init_hugepage: bind_area_to_memory_nodeset() failed (physical_device_id: 0 ch: 1). Hugepage allocation is not on NumaNode matching TT Device. Side-Effect is decreased Device->Host perf (Issue #893).
2024-05-30 15:00:50.885 | INFO     | SiliconDriver   - Software version 6.0.0, Ethernet FW version 6.3.0 (Device 0)
2024-05-30 15:00:50.886 | INFO     | SiliconDriver   - Software version 6.0.0, Ethernet FW version 6.3.0 (Device 1)
                  Metal | INFO     | Initializing device 0. Program cache is NOT enabled
                  Metal | INFO     | AI CLK for device 0 is:   1000 MHz
                     Op | WARNING  | reduce scalar doesn't support fp32_dest_acc_en! TEST BRANCH
           BuildKernels | INFO     | trisc1 compile failure -- cmd: cd /home/ubuntu/tt-metal/built/2304/kernels/moreh_bias_backward_single_core_hw/5265195927809415689/trisc1/ && /home/ubuntu/tt-metal/tt_metal/third_party/sfpi/compiler/bin/riscv32-unknown-elf-g++ -mwormhole -march=rv32imw -mtune=rvtt-b1 -mabi=ilp32 -std=c++17 -flto -ffast-math -fno-use-cxa-atexit -fno-exceptions -Wall -Werror -Wno-unknown-pragmas -Wno-error=multistatement-macros -Wno-error=parentheses -Wno-error=unused-but-set-variable -Wno-unused-variable -O3 -DARCH_WORMHOLE -DTENSIX_FIRMWARE -DLOCAL_MEM_EN=0 -DUCK_CHLKC_MATH -DNAMESPACE=chlkc_math -DCOMPILE_FOR_TRISC=1 -DKERNEL_BUILD -I. -I.. -I/home/ubuntu/tt-metal/ -I/home/ubuntu/tt-metal/tt_metal -I/home/ubuntu/tt-metal/tt_metal/include -I/home/ubuntu/tt-metal/tt_metal/hw/inc -I/home/ubuntu/tt-metal/tt_metal/hw/inc/debug -I/home/ubuntu/tt-metal/tt_metal/hw/inc/wormhole -I/home/ubuntu/tt-metal/tt_metal/hw/inc/wormhole/wormhole_b0_defines -I/home/ubuntu/tt-metal/tt_metal/hw/inc/wormhole/noc -I/home/ubuntu/tt-metal/tt_metal/third_party/umd/device/wormhole_b0 -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/common -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_io -I/home/ubuntu/tt-metal/tt_metal/third_party/tt_llk_wormhole_b0/common/inc -I/home/ubuntu/tt-metal/tt_metal/third_party/tt_llk_wormhole_b0/llk_lib -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/inc -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/common -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_io -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api -I/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu -I/home/ubuntu/tt-metal/tt_metal/third_party/sfpi/include -I/home/ubuntu/tt-metal/tt_metal/hw/firmware/src -I/home/ubuntu/tt-metal/tt_metal/third_party/tt_llk_wormhole_b0/llk_lib -c -o trisck.o /home/ubuntu/tt-metal/tt_metal/hw/firmware/src/trisck.cc
In file included from /home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_math_reduce_api.h:7,
                 from /home/ubuntu/tt-metal/tt_metal/include/compute_kernel_api.h:23,
                 from /home/ubuntu/tt-metal/tt_eager/tt_dnn/kernels/compute/moreh_common.hpp:19,
                 from /home/ubuntu/tt-metal/tt_eager/tt_dnn/op_library/moreh_linear_backward/kernels/moreh_bias_backward_single_core_hw.cpp:5,
                 from ../chlkc_math.cpp:3,
                 from /home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/common/chlkc_list.h:20,
                 from /home/ubuntu/tt-metal/tt_metal/hw/firmware/src/trisck.cc:14:
/home/ubuntu/tt-metal/tt_metal/third_party/tt_llk_wormhole_b0/llk_lib/llk_math_reduce.h: In instantiation of 'void _llk_math_reduce_(uint, bool, uint) [with ckernel::PoolType type = ckernel::SUM; ckernel::ReduceDim dim = ckernel::REDUCE_SCALAR; int MATH_FIDELITY_DESC = 4; bool is_fp32_dest_acc_en = true; bool is_int_fpu_en = false; uint = long unsigned int]':
/home/ubuntu/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_math_reduce_api.h:20:90:   required from 'void llk_math_reduce(uint, uint) [with ckernel::PoolType type = ckernel::SUM; ckernel::ReduceDim dim = ckernel::REDUCE_SCALAR; int num_fidelity_phases = 4; bool is_fp32_dest_acc_en = true; bool is_int_fpu_en = false; uint = long unsigned int]'
/home/ubuntu/tt-metal/tt_metal/include/compute_kernel_api/reduce.h:103:5:   required from 'void ckernel::reduce_tile(uint32_t, uint32_t, uint32_t, uint32_t, uint32_t) [with ckernel::PoolType reduce_type = ckernel::SUM; ckernel::ReduceDim reduce_dim = ckernel::REDUCE_SCALAR; uint32_t = long unsigned int]'
/home/ubuntu/tt-metal/tt_eager/tt_dnn/kernels/compute/moreh_common.hpp:490:55:   required from here
/home/ubuntu/tt-metal/tt_metal/third_party/tt_llk_wormhole_b0/llk_lib/llk_math_reduce.h:195:23: error: static assertion failed
  195 |         static_assert(!is_fp32_dest_acc_en);
      |                       ^~~~~~~~~~~~~~~~~~~~
                 Always | FATAL    | trisc1 build failed
                 Always | FATAL    | Failed to generate binaries for moreh_bias_backward_single_core_hw TT_THROW @ ../tt_metal/jit_build/build.cpp:425: tt::exception
info:
trisc1 build failed
backtrace:
```

Please let me know if you have nay questions or need further assistance.

Best regards,

Dongjin.

**Comment by ttmtrajkovic** [comment]
Thank you @dongjin-na. Will address this tomorrow

**Comment by ttmtrajkovic** [comment]
Hey @dongjin-na, I've reviewed the code and the assert seems incorrect as ZEROACC instruction should work for fp32 mode. I've commented out the assert and the test has passed. 
I can go ahead and push this, but could you please just try this on your end to confirm that the test is indeed passing (i.e. doing something meaningful) as opposed to a false pass?

```
     } else if constexpr (dim == ReduceDim::REDUCE_SCALAR) {
         //fp32 dest unsupported with reduce scalar, must fix zeroacc
-        static_assert(!is_fp32_dest_acc_en);
+        // static_assert(!is_fp32_dest_acc_en);
         for (int tile = 0; tile < 3; tile++) {
             // Wait and pool
             if constexpr (type == PoolType::MAX) {
```


Milos

**Comment by dongjin-na** [comment]
Dear @ttmtrajkovic,
Thanks for the check.
I ran `test_moreh_linear_backward`, which has test cases of REDUCE_SCALAR, which were passed in fp32 mode.

@hschoi4448, are there any other test cases that use REDUCE_SCALAR in fp32 mode?

**Comment by ttmtrajkovic** [comment]
Changes are merged at https://github.com/tenstorrent/tt-metal/pull/9188

@dongjin-na , please close the issue if you see no other issues.
